### ISSUE TITLE
PBS not releasing dependent job

### ISSUE THREAD

**Issue by rispoli** [PBSPro/pbspro#1670]
Dear all,
PBS isn't releasing a job depending on another, when the jobs are submitted from a compute node rather than the master.

I generated a very simple "pre" and "post" job to reproduce the issue, "pre":
```#!/bin/bash
#PBS -N pre
#PBS -l walltime=01:00:00
#PBS -q workq
#PBS -j oe
#PBS -l select=1:ncpus=1:mpiprocs=1

echo "$(date)"
echo $HOSTNAME
sleep 1m
echo "$(date)"
```
post:
```#!/bin/bash
#PBS -N post
#PBS -l walltime=01:00:00
#PBS -q workq
#PBS -j oe
#PBS -l select=1:ncpus=1:mpiprocs=1

echo "$(date)"
echo $HOSTNAME
sleep 1m
echo "$(date)"
```

Submitted them with:
```
$ qsub dep_test_pre.sh
$ qsub -W depend=afterok:<JOB-ID-HERE> dep_test_post.sh
```

After 1m, once the "pre" job completes, PBS notices and signals for "post" to be released:
```
$ grep '18.ip-' /var/spool/pbs/server_logs/20200417
04/17/2020 17:19:52;0100;Server@ip-0a000205;Job;18.ip-0A000205;enqueuing into workq, state 2 hop 1
04/17/2020 17:19:52;0008;Server@ip-0a000205;Job;18.ip-0A000205;Job Queued at request of daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, owner = daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, job name = post, queue = workq
04/17/2020 17:20:00;0006;Server@ip-0a000205;Hook;Server@ip-0a000205;cycle_periodic_hook_place - Jobs: ['18.ip-0A000205']
04/17/2020 17:20:00;0006;Server@ip-0a000205;Hook;Server@ip-0a000205;cycle_periodic_hook_place - Cmd: ['/opt/pbs/bin/qstat', '-f', '-F', 'json', '18.ip-0A000205']
04/17/2020 17:20:00;0006;Server@ip-0a000205;Hook;Server@ip-0a000205;cycle_periodic_hook_place - Key: 18.ip-0A000205
04/17/2020 17:20:00;0006;Server@ip-0a000205;Hook;Server@ip-0a000205;cycle_periodic_hook_place - full cmd ['/opt/pbs/bin/qalter', u'-lselect=1:ncpus=1:mpiprocs=1', u'-lplace=scatter:group=group_id', u'18.ip-0A000205']
04/17/2020 17:20:00;0006;Server@ip-0a000205;Hook;Server@ip-0a000205;cycle_periodic_hook_place - Cmd: ['/opt/pbs/bin/qalter', u'-lselect=1:ncpus=1:mpiprocs=1', u'-lplace=scatter:group=group_id', u'18.ip-0A000205']
04/17/2020 17:20:00;0008;Server@ip-0a000205;Job;18.ip-0A000205;Job Modified at request of root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/17/2020 17:20:00;0006;Server@ip-0a000205;Hook;Server@ip-0a000205;cycle_periodic_hook_place - Release the hold on job 18.ip-0A000205
04/17/2020 17:20:00;0006;Server@ip-0a000205;Hook;Server@ip-0a000205;cycle_periodic_hook_place - Cmd: ['/opt/pbs/bin/qrls', '-h', 'o', u'18.ip-0A000205']
04/17/2020 17:20:00;0008;Server@ip-0a000205;Job;18.ip-0A000205;Holds o released at request of root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
```
but nothing ever happens as the job is still being held:
```
$ qstat -f 18
Job Id: 18.ip-0A000205
    Job_Name = post
    Job_Owner = daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.clou
        dapp.net
    job_state = H
    queue = workq
    server = ip-0A000205
    Checkpoint = u
    ctime = Fri Apr 17 17:19:52 2020
    depend = afterok:17.ip-0A000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.clou
        dapp.net@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.ne
        t
    Error_Path = ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.ne
        t:/shared/home/daniele/good_tests_compute/post.e18
    Hold_Types = s
    Join_Path = oe
    Keep_Files = n
    Mail_Points = a
    mtime = Fri Apr 17 17:20:00 2020
    Output_Path = ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.n
        et:/shared/home/daniele/good_tests_compute/post.o18
    Priority = 0
    qtime = Fri Apr 17 17:19:52 2020
    Rerunable = True
    Resource_List.mpiprocs = 1
    Resource_List.ncpus = 1
    Resource_List.nodect = 1
    Resource_List.place = scatter:group=group_id
    Resource_List.select = 1:ncpus=1:mpiprocs=1
    Resource_List.ungrouped = false
    Resource_List.walltime = 01:00:00
    substate = 22
    Variable_List = PBS_O_SYSTEM=Linux,PBS_O_SHELL=/bin/bash,
        PBS_O_HOME=/shared/home/daniele,PBS_O_LOGNAME=daniele,
        PBS_O_WORKDIR=/shared/home/daniele/good_tests_compute,
        PBS_O_LANG=C.UTF-8,
        PBS_O_PATH=/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/cycl
        e/jetpack/bin:/opt/ibutils/bin:/opt/pbs/bin:/shared/home/daniele/.local
        /bin:/shared/home/daniele/bin,PBS_O_MAIL=/var/spool/mail/daniele,
        PBS_O_QUEUE=workq,
        PBS_O_HOST=ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp
        .net
    Submit_arguments = -W depend=afterok:17 ../dep_test_post.sh
    project = _pbs_project_default
```

Please note this only fails when executing the `qsub` from a compute node, when submitting from the master all goes smoothly.
In order to be able to submit from the compute node `sudo qmgr -c "s s flatuid = true"` on the master has to be executed, even though the IDs and GIDs are the same for that user, might that be a factor?

Other details:
 - running on Azure cloud;
 - cluster created by CycleCloud;
 - PBS version: 18.1.4, but a customer has the same issue with a slightly older version;
 - CentOS 7.7;
 - SELinux set to Permissive.

Any suggestions on what it might be or how to troubleshoot further?

Thanks in advance for any help you might provide,
Daniele

**Comment by arungrover** [comment]
@rispoli can you please issue the command - "tracejob 18.ip-0A000205" and "tracejob 17.ip-0A000205" on the machine where PBS server is hosted?
It seems like the hold release message you are seeing is to remove Hold type 'o' which is typically put by an admin.
When PBS holds a job due to dependency, it puts a system hold on the job (type s) which I can still see in the qstat output of the dependent job.

**Comment by rispoli** [comment]
Dear @arungrover,
thanks for your quick answer, here's the output:
```
$ tracejob 17 -n 30

Job: 17.ip-0A000205

04/17/2020 17:19:21  S    enqueuing into workq, state 2 hop 1
04/17/2020 17:19:21  S    Job Queued at request of
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net,
                          owner =
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, job
                          name = pre, queue = workq
04/17/2020 17:19:30  L    Considering job to run
04/17/2020 17:19:30  S    Job Modified at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/17/2020 17:19:30  S    Holds so released at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/17/2020 17:19:30  S    Job Modified at request of
                          Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/17/2020 17:19:30  S    Job Run at request of
                          Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net on
                          exec_vnode (ip-0A000207:ncpus=1)
04/17/2020 17:19:30  L    Job run
04/17/2020 17:20:30  S    Obit received momhop:1 serverhop:1 state:4 substate:42
04/17/2020 17:20:30  S    Exit_status=0 resources_used.cpupercent=0 resources_used.cput=00:00:00
                          resources_used.mem=3484kb resources_used.ncpus=1
                          resources_used.vmem=336436kb resources_used.walltime=00:01:00
04/17/2020 17:20:30  S    dequeuing from workq, state 5
```
and:
```
$ tracejob 18 -n 30

Job: 18.ip-0A000205

04/17/2020 17:19:52  S    enqueuing into workq, state 2 hop 1
04/17/2020 17:19:52  S    Job Queued at request of
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net,
                          owner =
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, job
                          name = post, queue = workq
04/17/2020 17:20:00  S    Job Modified at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/17/2020 17:20:00  S    Holds o released at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
```

Wouldn't that confirm my suspicion that `sudo qmgr -c "s s flatuid = false"` might somehow be involved?
Are there other ways to enabling submission from compute nodes?

Regards,
Daniele

**Comment by arungrover** [comment]
Hi @rispoli Thanks for providing the logs. It appears that your suspicion is correct. The issue does look related to flatuid. Since you have flatuid as False, PBS to verify the user tries to ruserok library call to ensure the user from the remote machine is the same user. It seems like ruserok call is failing which is causing PBS to not release the dependency.
Can you please set flatuid to 1 or check user's .rhosts for right remote system names and permissions.

**Comment by rispoli** [comment]
Hi @arungrover,
if I try to submit with `flatuid = true` I get the following error preventing me from doing so:
```
$ qsub ../dep_test_pre.sh
qsub: Bad UID for job execution
```

Is there another way to allow user submissions from a compute node without incurring in the `ruserok` issue you describe?

Furthermore `.rhosts` isn't defined for that system, could we skip "messing about" with `flatuid` if we had that set up?
Unfortunately, though, that seems to be hostname-based, and hostnames in a cloud environment aren't guaranteed to be consistent or repeatable.
How would you approach that conoundrum?

**Comment by rispoli** [comment]
Just did another test with `flatuid` to its original setting, confirmed by the `Bad UID` error when trying to submit the first time, then edited an `.rhosts` manually to allow submission from that host.

The result is the same, both for an `afterok` and an `afterany` dependency:
```
$ qsub ../dep_test_pre.sh
30.ip-0A000205
$ qsub -W depend=afterok:30 ../dep_test_post.sh
31.ip-0A000205
$ qsub -W depend=afterany:30 ../dep_test_post.sh
32.ip-0A000205
```
which results in:
```
$ tracejob 30

Job: 30.ip-0A000205

04/20/2020 21:04:22  S    enqueuing into workq, state 2 hop 1
04/20/2020 21:04:22  S    Job Queued at request of
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net,
                          owner =
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, job
                          name = pre, queue = workq
04/20/2020 21:04:30  L    Considering job to run
04/20/2020 21:04:30  S    Job Modified at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/20/2020 21:04:30  S    Holds so released at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/20/2020 21:04:30  S    Job Modified at request of
                          Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/20/2020 21:04:30  S    Job Run at request of
                          Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net on
                          exec_vnode (ip-0A000207:ncpus=1)
04/20/2020 21:04:30  L    Job run
04/20/2020 21:05:31  S    Obit received momhop:1 serverhop:1 state:4 substate:42
04/20/2020 21:05:31  S    Exit_status=0 resources_used.cpupercent=0 resources_used.cput=00:00:00
                          resources_used.mem=3472kb resources_used.ncpus=1
                          resources_used.vmem=336436kb resources_used.walltime=00:01:00
04/20/2020 21:05:31  S    dequeuing from workq, state 5
$ tracejob 31

Job: 31.ip-0A000205

04/20/2020 21:04:31  S    enqueuing into workq, state 2 hop 1
04/20/2020 21:04:31  S    Job Queued at request of
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net,
                          owner =
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, job
                          name = post, queue = workq
04/20/2020 21:04:45  S    Job Modified at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/20/2020 21:04:45  S    Holds o released at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
$ tracejob 32

Job: 32.ip-0A000205

04/20/2020 21:04:37  S    enqueuing into workq, state 2 hop 1
04/20/2020 21:04:37  S    Job Queued at request of
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net,
                          owner =
                          daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, job
                          name = post, queue = workq
04/20/2020 21:04:45  S    Job Modified at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/20/2020 21:04:45  S    Holds o released at request of
                          root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
```
and, as an additional data point, on the compute node:
```
$ tracejob 30

Job: 30.ip-0A000205

04/20/2020 21:04:31  M    Started, pid = 7399
04/20/2020 21:05:31  M    task 00000001 terminated
04/20/2020 21:05:31  M    Terminated
04/20/2020 21:05:31  M    task 00000001 cput= 0:00:00
04/20/2020 21:05:31  M    kill_job
04/20/2020 21:05:31  M    ip-0A000207 cput= 0:00:00 mem=3472kb
04/20/2020 21:05:31  M    no active tasks
04/20/2020 21:05:31  M    Obit sent
04/20/2020 21:05:31  M    copy file request received
04/20/2020 21:05:31  M    staged 1 items out over 0:00:00
04/20/2020 21:05:31  M    no active tasks
04/20/2020 21:05:31  M    delete job request received
04/20/2020 21:05:31  M    kill_job
$ tracejob 31

tracejob: Couldn't find Job Id 31.ip-0A000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net in logs of past 1 day

$ tracejob 32

tracejob: Couldn't find Job Id 32.ip-0A000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net in logs of past 1 day
```

I can grant remote (SSH) access to this machine for further troubleshooting, if needed.

Thanks once more for your help,
Daniele

**Comment by arungrover** [comment]
@rispoli sorry about the delayed reply. I guess it would be better to debug the application on the machine. If you can, please share access details. My email is arun.grover@altair.com

**Comment by rispoli** [comment]
@arungrover email sent.

**Comment by arungrover** [comment]
Thanks @rispoli for providing access to the machines.
I think the problem is with how PBS_SERVER name is setup in the execution host from where you are submitting the jobs.
It points to a long name - ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
and mismatches with PBS_SERVER on the server (ip-0A000205)

Because of this mismatch, when the parent job ends and the server tries to find the dependent job, the job name string comparison fails and no dependency is released.
I changed pbs.conf file on one of the execution hosts to have consistent PBS_SERVER name and it worked -
```[daniele@ip-0A000205 ~]$ tracejob 54 55

Job: 54.ip-0A000205

04/23/2020 23:08:50  S    enqueuing into workq, state 2 hop 1
04/23/2020 23:08:50  S    Job Queued at request of daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, owner = daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net,
                          job name = pre, queue = workq
04/23/2020 23:08:56  L    Considering job to run
04/23/2020 23:08:56  S    Job Modified at request of root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/23/2020 23:08:56  S    Holds so released at request of root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/23/2020 23:08:56  S    Job Modified at request of Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/23/2020 23:08:56  S    Job Run at request of Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net on exec_vnode (ip-0A000207:ncpus=1)
04/23/2020 23:08:56  L    Job run
04/23/2020 23:09:56  S    Obit received momhop:1 serverhop:1 state:4 substate:42
04/23/2020 23:09:56  S    Exit_status=0 resources_used.cpupercent=0 resources_used.cput=00:00:00 resources_used.mem=3476kb resources_used.ncpus=1 resources_used.vmem=336436kb
                          resources_used.walltime=00:01:00
04/23/2020 23:09:56  S    dequeuing from workq, state 5

Job: 55.ip-0A000205

04/23/2020 23:09:17  S    enqueuing into workq, state 2 hop 1
04/23/2020 23:09:17  S    Job Queued at request of daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net, owner = daniele@ip-0a000207.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net,
                          job name = post, queue = workq
04/23/2020 23:09:26  S    Job Modified at request of root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/23/2020 23:09:26  S    Holds o released at request of root@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/23/2020 23:09:56  L    Considering job to run
04/23/2020 23:09:56  S    Dependency on job 54.ip-0A000205 released.
04/23/2020 23:09:56  S    Job Modified at request of Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/23/2020 23:09:56  L    Not enough free nodes available
04/23/2020 23:09:57  L    Considering job to run
04/23/2020 23:09:57  S    Job Modified at request of Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net
04/23/2020 23:09:57  S    Job Run at request of Scheduler@ip-0a000205.htpsa2udrriuvayrbdpyqjz5ae.ax.internal.cloudapp.net on exec_vnode (ip-0A000207:ncpus=1)
04/23/2020 23:09:57  L    Job run
[daniele@ip-0A000205 ~]$
```
I'd suggest you keep PBS_SERVER name consistent across all your machines and that should solve this problem.

**Comment by rispoli** [comment]
Thanks @arungrover. We replicated the issue internally and are investigating ways to address it.
Closing this ticket for the time being.
